packages/ai-chat/src/common/chat-model.ts

```diff
@@ -0,0 +1,753 @@
+    readonly response: ChatResponse;
+    readonly isComplete: boolean;
+    readonly isCanceled: boolean;
+    readonly isError: boolean;
+    readonly agentId?: string
+    cancel(): void;
+    error(error: Error): void;
+    readonly errorObject?: Error;
+
+}
+
+/**********************
+ * Implementations
+ **********************/
+
+export class ChatModelImpl implements ChatModel {
+    protected readonly _onDidChangeEmitter = new Emitter<ChatChangeEvent>();
+    onDidChange: Event<ChatChangeEvent> = this._onDidChangeEmitter.event;
+
+    protected _requests: ChatRequestModelImpl[];
+    protected _id: string;
+
+    constructor(public readonly location = ChatAgentLocation.Panel) {
+        // TODO accept serialized data as a parameter to restore a previously saved ChatModel
+        this._requests = [];
+        this._id = generateUuid();
+    }
+
+    getRequests(): ChatRequestModelImpl[] {
+        return this._requests;
+    }
+
+    get id(): string {
+        return this._id;
+    }
+
+    addRequest(parsedChatRequest: ParsedChatRequest, agentId?: string): ChatRequestModelImpl {
+        const requestModel = new ChatRequestModelImpl(this, parsedChatRequest, agentId);
+        this._requests.push(requestModel);
+        this._onDidChangeEmitter.fire({
+            kind: 'addRequest',
+            request: requestModel,
+        });
+        return requestModel;
+    }
+
+    isEmpty(): boolean {
+        return this._requests.length === 0;
+    }
+}
+
+export class ChatRequestModelImpl implements ChatRequestModel {
+    protected _id: string;
+    protected _session: ChatModel;
+    protected _request: ChatRequest;
+    protected _response: ChatResponseModelImpl;
+    protected _agentId?: string;
+
+    constructor(session: ChatModel, public readonly message: ParsedChatRequest, agentId?: string) {
+        // TODO accept serialized data as a parameter to restore a previously saved ChatRequestModel
+        this._request = message.request;
+        this._id = generateUuid();
+        this._session = session;
+        this._response = new ChatResponseModelImpl(this._id, agentId);
+        this._agentId = agentId;
+    }
+
+    get id(): string {
+        return this._id;
+    }
+
+    get session(): ChatModel {
+        return this._session;
+    }
+
+    get request(): ChatRequest {
+        return this._request;
+    }
+
+    get response(): ChatResponseModelImpl {
+        return this._response;
+    }
+
+    get agentId(): string | undefined {
+        return this._agentId;
+    }
+}
+
+export class ErrorResponseContentImpl implements ErrorResponseContent {
+    kind: 'error' = 'error';
+    protected _error: Error;
+    constructor(error: Error) {
+        this._error = error;
+    }
+    get error(): Error {
+        return this._error;
+    }
+    asString(): string | undefined {
+        return undefined;
+    }
+}
+
+export class TextChatResponseContentImpl implements TextChatResponseContent {
+    kind: 'text' = 'text';
```

Comment:
No this will not work as `Typescript` will determine the value to be of type `string`, however in the `TextChatResponseContent` the type is `'text'`. Therefore we need to repeat the type which let's Typescript realize that they are compatible.

Reply:
Did you try it? I believe TS figures this out just fine.

Reply:
>Property 'kind' in type 'TextChatResponseContentImpl' is not assignable to the same property in base type 'TextChatResponseContent'.
  Type 'string' is not assignable to type '"text"'.ts(2416)

Reply:
Is it really compiling just fine or is it just the IDE not complaining? For me both complain. In case your IDE uses a different Typescript version then you might not get an error there.

i.e. `yarn` produces
```
packages/ai-chat/src/common/chat-model.ts:391:5 - error TS2416: Property 'kind' in type 'TextChatResponseContentImpl' is not assignable to the same property in base type 'TextChatResponseContent'.
  Type 'string' is not assignable to type '"text"'.

391     kind = 'text';
        ~~~~


Found 1 error.

error Command failed with exit code 1.
```